src/control/cmd/dmg/pool.go

```diff
@@ -184,50 +158,27 @@ func (d *PoolDestroyCmd) Execute(args []string) error {
 // PoolQueryCmd is the struct representing the command to destroy a DAOS pool.
 type PoolQueryCmd struct {
 	logCmd
-	connectedCmd
+	ctlClientCmd
 	UUID string `long:"pool" required:"1" description:"UUID of DAOS pool to query"`
 }
 
 // Execute is run when PoolQueryCmd subcommand is activated
 func (c *PoolQueryCmd) Execute(args []string) error {
-	req := client.PoolQueryReq{
+	req := &control.PoolQueryReq{
 		UUID: c.UUID,
 	}
+	req.SetHostList(c.hostlist)
 
-	resp, err := c.conns.PoolQuery(req)
+	ctx := context.Background()
+	resp, err := control.PoolQuery(ctx, c.ctlClient, req)
 	if err != nil {
 		return errors.Wrap(err, "pool query failed")
 	}
 
-	// Maintain output compability with the `daos pool query` output.
 	var bld strings.Builder
-	fmt.Fprintf(&bld, "Pool %s, ntarget=%d, disabled=%d\n",
-		resp.UUID, resp.TotalTargets, resp.DisabledTargets)
-	bld.WriteString("Pool space info:\n")
-	fmt.Fprintf(&bld, "- Target(VOS) count:%d\n", resp.ActiveTargets)
-	if resp.Scm != nil {
-		bld.WriteString("- SCM:\n")
-		fmt.Fprintf(&bld, "  Total size: %s\n", humanize.Bytes(resp.Scm.Total))
-		fmt.Fprintf(&bld, "  Free: %s, min:%s, max:%s, mean:%s\n",
-			humanize.Bytes(resp.Scm.Free), humanize.Bytes(resp.Scm.Min),
-			humanize.Bytes(resp.Scm.Max), humanize.Bytes(resp.Scm.Mean))
-	}
-	if resp.Nvme != nil {
-		bld.WriteString("- NVMe:\n")
-		fmt.Fprintf(&bld, "  Total size: %s\n", humanize.Bytes(resp.Nvme.Total))
-		fmt.Fprintf(&bld, "  Free: %s, min:%s, max:%s, mean:%s\n",
-			humanize.Bytes(resp.Nvme.Free), humanize.Bytes(resp.Nvme.Min),
-			humanize.Bytes(resp.Nvme.Max), humanize.Bytes(resp.Nvme.Mean))
-	}
-	if resp.Rebuild != nil {
-		if resp.Rebuild.Status == 0 {
-			fmt.Fprintf(&bld, "Rebuild %s, %d objs, %d recs\n",
-				resp.Rebuild.State, resp.Rebuild.Objects, resp.Rebuild.Records)
-		} else {
-			fmt.Fprintf(&bld, "Rebuild failed, rc=%d, status=%d", resp.Status, resp.Rebuild.Status)
-		}
+	if err := control.PrintPoolQueryResponse(resp, &bld); err != nil {
```

Comment:
The pretty printer functions are intended to be optional conveniences, and they're not tied into any of the core logic. We can't assume that dmg will be the only user of the API or the only user that will want to pretty-print the structures. I agree that we shouldn't bring dmg specifics into the control API, but I don't think there's anything specific to dmg in these pretty printers -- they're just a decent and convenient way of representing information-dense API responses for humans.

If we pull these out into dmg, then anyone else wanting to use them will have to copy them out of dmg because it's not possible to import from a main package. Either that, or they'll have to write their own, and that seems pointlessly hostile as well.

Reply:
My thought wasn't that dmg is the only consumer, more that different tools will have different requirements for output display. DMG requirements are what dictate the changes that will be accepted for these pretty printers. It's that coupling that makes me uncomfortable with having it here.

Reply:
Sure, I can see the argument that dmg's display requirements shouldn't be driving changes to control API code. What I'm hoping to avoid is exposing too much of our internal APIs beyond the control API... In other words, most users (including dmg), will only import `github.com/daos-stack/daos/src/control/lib/control`, and not random things like `control/common` (which is turning into a huge grab bag of unrelated functionality).

What if we put the pretty-printers under `lib/control/pretty` ? Then they're not part of the top-level API surface, but they're still available for API users. They wouldn't necessarily need to change in response to dmg's output requirements. If dmg needs something that the standard pretty-printers don't provide, then dmg is free to implement its own thing.

I don't plan to add pretty-printers for each response type. The thought was that complex responses with a common structure (e.g. HostErrors/HostStorage maps) could all benefit from standard pretty-printers that do a better job of printing scale-friendly output.

Reply:
On that topic, I can pull the PoolQueryResp pretty-printing back out -- I don't feel strongly about keeping that one in. Maybe keeping that in dmg will make the other printers seem less tied to dmg?

Reply:
lib/control/pretty sounds good to me

Reply:
I'm going to pull them all back out under dmg/pretty. That way they're distinct from the rest of the dmg machinery, and if they're truly reusable they can be imported in a pinch or we can move them somewhere else. For the moment I'll just assume that any callers of the control API that want some representation other than "%+v" will dump the responses to JSON or implement their own way of dealing with the responses.

Reply:
sounds good

Reply:
Hmm. Pulling the pretty printers out into a separate package is turning out to be not at all a clean effort. In order to generate responses for testing the printer outputs, I was using unexported mock helpers to create complex synthetic responses for various test cases. In order to achieve the same level of test coverage in a separate package, I have to either promote all of the mock helpers to exported functions, bloating the public control API surface, or I need to effectively duplicate the helpers over in dmg/pretty. That also seems wrong.

I still agree that one-off pretty printers like the pool query response printer should probably live in dmg/pretty, but I'm having a hard time justifying the amount of work it'd take to extract the other core response printers, given how tightly-coupled they are to the control package.

I do have a longer-term goal of somehow generating synthetic API responses as part of a integration testing/benchmarking framework, but I don't yet know exactly what that will look like. Unless there are strong objections, I'd like to move forward with what we have now and revisit this issue later.

Reply:
I'm fine with leaving them as-is if it's a bear to move them. It's a little odd to me that we need to use complex mock setup functions to initialize some structures to be formatted for print, but that's something we can look at in the future, too.